coralproject/talk
# [next] Story Mutations

## What does this PR do?

This introduces some breaking changes to the actions collection, which requires the database to be dropped to support.

### Refactors

- Renamed all delete* to remove* to avoid reserved keyword collision

### Added Mutations

```graphql
  """
  createStory will create the provided Story.
  """
  createStory(input: CreateStoryInput!): CreateStoryPayload
    @auth(roles: [ADMIN])

  """
  updateStory will update the given Story.
  """
  updateStory(input: UpdateStoryInput!): UpdateStoryPayload
    @auth(roles: [ADMIN])

  """
  mergeStories will merge two stories together, merging their comment streams.
  This operation is irreversible.
  """
  mergeStories(input: MergeStoriesInput!): MergeStoriesPayload
    @auth(roles: [ADMIN])

  """
  removeStory will remove the given Story.
  """
  removeStory(input: RemoveStoryInput!): RemoveStoryPayload
    @auth(roles: [ADMIN])

  """
  scrapeStory will scrape the given Story and update the scraped metadata.
  """
  scrapeStory(input: ScrapeStoryInput!): ScrapeStoryPayload
    @auth(roles: [ADMIN, MODERATOR])
```

## How do I test this PR?

Ensure that https://coralproject.net/ is in your list of allowed domains. Grab a valid admin token from localstorage before these tests, and run:

```bash
export TOKEN=<that token>
```

So that the following curl requests will work!

### Create a Story:

```bash
curl --request POST \
  --url http://localhost:3000/api/tenant/graphql \
  --header "authorization: Bearer $TOKEN" \
  --header 'content-type: application/json' \
  --data '{"query":"mutation ($story: CreateStory!) {\n  createStory(input: {story: $story, clientMutationId: \"\"}) {\n    story {\n      id\n      url\n      metadata {\n        title\n        author\n        description\n        image\n        publishedAt\n        modifiedAt\n        section\n      }\n      scrapedAt\n    }\n  }\n}\n","variables":{"story":{"id":"a1e3a709-e82d-4d48-a4c2-b24ec2991721","url":"https://coralproject.net/blog/happy-hacktoberfest/"}}}'
```

Notice that it responds with scraped asset data, that's because we didn't provide it, so the default was to scrape it!

Try now to create a story with some metadata:

```bash
curl --request POST \
  --url http://localhost:3000/api/tenant/graphql \
  --header "authorization: Bearer $TOKEN" \
  --header 'content-type: application/json' \
  --data '{"query":"mutation ($story: CreateStory!) {\n  createStory(input: {story: $story, clientMutationId: \"\"}) {\n    story {\n      id\n      url\n      metadata {\n        title\n        author\n        description\n        image\n        publishedAt\n        modifiedAt\n        section\n      }\n      scrapedAt\n    }\n  }\n}\n","variables":{"story":{"id":"c1e3a709-e82d-4d48-a4c2-b24ec2991721","url":"https://coralproject.net/blog/survey-marginalized-voices/","metadata":{"title":"This is a title"}}}}'
```

Note that the rest of the metadata was not filled in, that's because we provided metadata!

### Rescraping a Story

```bash
curl --request POST \
  --url http://localhost:3000/api/tenant/graphql \
  --header "authorization: Bearer $TOKEN" \
  --header 'content-type: application/json' \
  --data '{"query":"mutation {\n  scrapeStory(input: {id: \"c1e3a709-e82d-4d48-a4c2-b24ec2991721\", clientMutationId: \"\"}) {\n    story {\n\t\t\turl\n      metadata {\n        title\n        author\n        description\n        image\n        publishedAt\n        modifiedAt\n        section\n\t\t\t}\n\t\t\tscrapedAt\n    }\n  }\n}\n"}'
```

Notice the scraped data!

### Updating a Story

```bash
curl --request POST \
  --url http://localhost:3000/api/tenant/graphql \
  --header "authorization: Bearer $TOKEN" \
  --header 'content-type: application/json' \
  --data '{"query":"mutation ($id: ID!, $story: UpdateStory!) {\n  updateStory(input: {id: $id, story: $story, clientMutationId: \"\"}) {\n    story {\n      id\n      url\n      metadata {\n        title\n        author\n        description\n        image\n        publishedAt\n        modifiedAt\n        section\n      }\n      scrapedAt\n    }\n  }\n}\n","variables":{"id":"c1e3a709-e82d-4d48-a4c2-b24ec2991721","story":{"metadata":{"title":"Updated Story Title"}}}}'
```

Notice that we updated the title, but the rest of the data is unchanged from before!

### Merging a story

Using the development command: `npm run watch client`, hit the "story" and "story with article" to generate two stories that we will use to merge together. Leave a comment on both of them, and "react" to both of those comments.

Looking at the GraphQL calls, grab the ID's for both of the stories.

Execute the following GraphQL call:

```graphql
mutation ($sourceIDs: [ID!]!, $destinationID: ID!) {
  mergeStories(input: {destinationID: $destinationID, sourceIDs: $sourceIDs, clientMutationId: ""}) {
    story {
      id
      url
      metadata {
        title
        author
        description
        image
        publishedAt
        modifiedAt
        section
      }
      scrapedAt
    }
  }
}
```

Substituting the variables with:

```json
{
	"destinationID": "<first page id>",
	"sourceIDs": [
		"<second page id>"
	]
}
```

Now visit the story of the "first" page, notice how it contains all the comments + reactions of the other story as well! Navigate back to the first story, notice how there is nothing on it now!

### Removing a story

Grab the "first page id" from above, and the new "second page id" after you visited it after you merged them before.

Try to remove both comments using the following graph query:

```graphql
mutation ($id: ID!) {
  removeStory(input: {id: $id, includeComments: false, clientMutationId: ""}) {
    story {
			id
      url
      metadata {
        title
        author
        description
        image
        publishedAt
        modifiedAt
        section
      }
      scrapedAt
    }
  }
}
```

Notice how the first one returns an error, and the second one does not? Try running this one instead for the first query:

```graphql
mutation ($id: ID!) {
  removeStory(input: {id: $id, includeComments: true, clientMutationId: ""}) {
    story {
			id
      url
      metadata {
        title
        author
        description
        image
        publishedAt
        modifiedAt
        section
      }
      scrapedAt
    }
  }
}
```

Note that `includeComments` is now `true`! Also note that there is no error. Navigate back to both pages, and see no comments or reactions on either of them.
